src/main/java/org/folio/circulation/services/AllowedServicePointsService.java

```diff
@@ -96,111 +118,123 @@ private CompletableFuture<Result<RequestPolicy>> lookupRequestPolicy(Item item,
 
-    log.debug("mapToAllowedServicePoints:: parameters: allowedTypes={}, servicePointsIds={}",
-      () -> asJson(allowedTypes), () -> asJson(servicePointIds));
+    log.debug("groupAllowedServicePointsByRequestType:: parameters allowedTypes: {}, " +
+        "servicePointsIds: {}, allowedServicePointsInPolicy: {}", () -> asJson(allowedTypes),
+      () -> asJson(allowedServicePoints), () -> allowedServicePointsInPolicy);
 
-    Map<RequestType, Set<String>> allowedServicePoints = new EnumMap<>(RequestType.class);
-    allowedTypes.forEach(requestType -> allowedServicePoints.put(requestType, servicePointIds));
-    log.info("fetchAllowedServicePoints:: result: {}", () -> asJson(allowedServicePoints));
-
-    return allowedServicePoints;
-  }
+    Map<RequestType, Set<AllowedServicePoint>> groupedAllowedServicePoints =
+      new EnumMap<>(RequestType.class);
+    for (Map.Entry<RequestType, Set<String>> entry : allowedServicePointsInPolicy.entrySet()) {
+      RequestType requestType = entry.getKey();
+      Set<String> servicePointIds = entry.getValue();
 
-  private Map<RequestType, Set<String>> includeOnlyExistingServicePoints(
-    Map<RequestType, Set<String>> allowedServicePointsInPolicy, Set<String> relevantSpIds,
-    RequestPolicy requestPolicy) {
-
-    log.debug("includeOnlyExistingServicePoints:: parameters: allowedServicePointsInPolicy={}, " +
-      "relevantSpIds={}, requestPolicy={}", () -> asJson(allowedServicePointsInPolicy),
-      () -> asJson(relevantSpIds), () -> asJson(requestPolicy));
-
-    final Map<RequestType, Set<String>> allowedServicePoints = new EnumMap<>(RequestType.class);
-    allowedServicePointsInPolicy.forEach((requestType, uuidSet) -> {
-      if (requestPolicy.allowsType(requestType)) {
-        log.info("includeOnlyExistingServicePoints:: requestType={} is allowed",
-          requestType.getValue());
-        Set<String> filteredIds = uuidSet.stream()
-          .filter(relevantSpIds::contains)
+      if (allowedTypes.contains(requestType) && !servicePointIds.isEmpty()) {
+        Set<AllowedServicePoint> allowedServicePointsForType = allowedServicePoints.stream()
+          .filter(allowedServicePoint -> servicePointIds.contains(allowedServicePoint.getId()))
           .collect(Collectors.toSet());
-        if (!filteredIds.isEmpty()) {
-          allowedServicePoints.put(requestType, filteredIds);
+
+        if (!allowedServicePointsForType.isEmpty()) {
+          groupedAllowedServicePoints.put(requestType, allowedServicePointsForType);
         }
       }
-    });
+    }
 
-    return allowedServicePoints;
+    return groupedAllowedServicePoints;
   }
 
-  private CompletableFuture<Result<Map<RequestType, Set<String>>>> getAllowedServicePointsForInstance(
-    AllowedServicePointsRequest request) {
+  private CompletableFuture<Result<Map<RequestType, Set<AllowedServicePoint>>>>
+  getAllowedServicePointsForInstance(AllowedServicePointsRequest request) {
```

Comment:
can we mode this method up closer to where it is called?

Reply:
done